openedx/core/djangoapps/util/management/commands/run_aggregator.py

```diff
@@ -0,0 +1,272 @@
+
+        # Other listeners are connected so we time the handler alone later.
+        from completion_aggregator.signals import course_published_handler
+        SignalHandler.course_published.disconnect(course_published_handler)
+
+        vertical = self.course.get_children()[-1].get_children()[-1].get_children()[-1]
+        self._complete_blocks_for_users(vertical.get_children(), self.users)
+        self._assert_vertical_completion_for_all_users(vertical, 1.0)
+        with self.store.branch_setting(ModuleStoreEnum.Branch.draft_preferred, self.course.id):
+            with self.store.bulk_operations(self.course.id):
+                self._create_block(parent=vertical, category='html')
+                self.store.publish(vertical.location, ModuleStoreEnum.UserID.test)
+
+        reset_queries()
+        time_taken = self._time_handler(course_published_handler, course_key=self.course.id)
+        self._copy_executed_queries()
+
+        self._print_results_header(u"test_course_published_handler_when_block_is_added", time_taken=time_taken)
+        self._assert_vertical_completion_for_all_users(
+            vertical, self.course_breadth[3] / (self.course_breadth[3] + 1.0)
+        )
+        self._print_results_footer()
+
+    def test_item_deleted_handler_when_block_is_deleted(self):
+
+        # Other listeners are connected so we time the handler alone later.
+        from completion_aggregator.signals import item_deleted_handler
+        SignalHandler.course_published.disconnect(item_deleted_handler)
+
+        vertical = self.course.get_children()[-1].get_children()[-1].get_children()[-1]
+        self._complete_blocks_for_users(vertical.get_children()[1:], self.users)
+        self._assert_vertical_completion_for_all_users(
+            vertical, (self.course_breadth[3] - 1.0) / self.course_breadth[3]
+        )
+        block = vertical.get_children()[0]
+        with self.store.branch_setting(ModuleStoreEnum.Branch.draft_preferred, self.course.id):
+            self.store.delete_item(block.location, ModuleStoreEnum.UserID.test)
+
+        reset_queries()
+        time_taken = self._time_handler(item_deleted_handler, usage_key=block.location, user_id=None)
+        self._copy_executed_queries()
+
+        self._print_results_header(u"test_item_deleted_handler_when_block_is_deleted", time_taken=time_taken)
+        self._assert_vertical_completion_for_all_users(vertical, 1.0)
+        self._print_results_footer()
+
+    def test_individual_block_completions(self):
+
+        from completion_aggregator.models import Aggregator
+
+        times_taken = []
+
+        for user in self.users:
+            user.next_block_index_to_complete = 0
+
+        users = list(self.users)
+
+        reset_queries()
+
+        for __ in range(self.completions_count):
+            random_user = random.choice(users)
+            next_block = self.blocks[random_user.next_block_index_to_complete]
```

Comment:
Just to clarify the intention here:  

Blocks will be completed by random users, in course order, until the total number of requested completions (completions_count) is reached, and if any user completes all the blocks in the course, they will be removed from selection for the remainder of the run.

Is that correct?

Reply:
Yes, that is correct.